ci/main.go

```diff
@@ -1,42 +1,367 @@
+		Include: []string{
+			// go source
+			"**/*.go",
+
+			// modules
+			"**/go.mod",
+			"**/go.sum",
+
+			// embedded files
+			"**/*.tmpl",
+			"**/*.ts.gtpl",
+			"**/*.graphqls",
+			"**/*.graphql",
+
+			// misc
+			".golangci.yml",
+			"**/README.md", // needed for examples test
+		},
+	})
+}
+
+func goBase() *Container {
+	repo := repositoryGoCodeOnly()
+
+	return dag.Container().
+		From(fmt.Sprintf("golang:%s-alpine%s", goVersion, alpineVersion)).
+		// gcc is needed to run go test -race https://github.com/golang/go/issues/9918 (???)
+		WithExec([]string{"apk", "add", "build-base"}).
+		WithEnvVariable("CGO_ENABLED", "0").
+		// adding the git CLI to inject vcs info into the go binaries
+		WithExec([]string{"apk", "add", "git"}).
+		WithWorkdir("/app").
+		// run `go mod download` with only go.mod files (re-run only if mod files have changed)
+		WithDirectory("/app", repo, ContainerWithDirectoryOpts{
+			Include: []string{"**/go.mod", "**/go.sum"},
+		}).
+		WithMountedCache("/go/pkg/mod", dag.CacheVolume("go-mod")).
+		WithExec([]string{"go", "mod", "download"}).
+		// run `go build` with all source
+		WithMountedDirectory("/app", repo).
+		// include a cache for go build
+		WithMountedCache("/root/.cache/go-build", dag.CacheVolume("go-build"))
+}
+
+func runcBin() *File {
+	return dag.HTTP(fmt.Sprintf(
+		"https://github.com/opencontainers/runc/releases/download/%s/runc.%s",
+		runcVersion,
+		runtime.GOARCH,
+	))
+}
+
+func buildctlBin() *File {
+	return goBase().
+		WithEnvVariable("GOOS", "linux").
+		WithEnvVariable("GOARCH", runtime.GOARCH).
+		WithExec([]string{
+			"go", "build",
+			"-o", "./bin/buildctl",
+			"-ldflags", "-s -w",
+			"github.com/moby/buildkit/cmd/buildctl",
+		}).
+		File("./bin/buildctl")
+}
+
+func shimBin() *File {
+	return goBase().
+		WithEnvVariable("GOOS", "linux").
+		WithEnvVariable("GOARCH", runtime.GOARCH).
+		WithExec([]string{
+			"go", "build",
+			"-o", "./bin/" + shimBinName,
+			"-ldflags", "-s -w",
+			"/app/cmd/shim",
+		}).
+		File("./bin/" + shimBinName)
+}
+
+func engineBin(version string) *File {
+	buildArgs := []string{
+		"go", "build",
+		"-o", "./bin/" + engineBinName,
+		"-ldflags",
+	}
+	ldflags := []string{"-s", "-w"}
+	if version != "" {
+		ldflags = append(ldflags, "-X", "github.com/dagger/dagger/engine.Version="+version)
 	}
-	return CITargets{
-		SrcDir: srcDir,
-	}, nil
+	buildArgs = append(buildArgs, strings.Join(ldflags, " "))
+	buildArgs = append(buildArgs, "/app/cmd/engine")
+	return goBase().
+		WithEnvVariable("GOOS", "linux").
+		WithEnvVariable("GOARCH", runtime.GOARCH).
+		WithExec(buildArgs).
+		File("./bin/" + engineBinName)
 }
 
-type CITargets struct {
-	SrcDir *dagger.Directory
+func qemuBins() *Directory {
+	return dag.Container().
+		From(qemuBinImage).
+		Rootfs()
 }
 
-// Dagger SDK targets
-func (t CITargets) SDK(ctx dagger.Context) (SDKTargets, error) {
-	return SDKTargets(t), nil
+func cniPlugins() *Directory {
```

Comment:
backporting the recent changes on main to build this stuff from source is on my TODO list